### PR TITLE
WIP: config/ingress: Add required route annotations API

### PR DESCRIPTION
**Pull request by Miciah**
* `config/v1/types_ingress.go` (`IngressSpec`): Add new `RequiredRouteAnnotations` field with type slice of `RequiredRouteAnnotations`.
(`RequiredRouteAnnotations`): New type.  Define annotations that are required to be set on newly created routes.
(`RequiredRouteAnnotationsPolicy`): New type.  Define policy for whether the user may provide arbitrary values for required annotations.
* `config/v1/0000_10_config-operator_01_ingress.crd.yaml`:
* `config/v1/zz_generated.deepcopy.go`:
* `config/v1/zz_generated.swagger_doc_generated.go`: Regenerate.

### CODE REVIEW

File: config/v1/types_ingress.go [hunk 1]
```diff
@@ -172,6 +180,98 @@ type ComponentRouteStatus struct {
...
+	// where the key specifies a required annotation key and the value
+	// expresses a suggested annotation value.  If a route specifies has an
+	// annotation with each of the required annotation keys, the route is
+	// admitted.  If the route is missing any required annotation, the route
+	// is rejected with a message indicating which annotation key is missing
+	// and what the suggested annotation value is.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	RequiredAnnotations map[string]string `json:"requiredAnnotations"`
```
Comment (deads2k) [inline config/v1/types_ingress.go:255]: what other required annotations do we have?

File: config/v1/types_ingress.go [hunk 2]
```diff
@@ -172,6 +180,98 @@ type ComponentRouteStatus struct {
 	RelatedObjects []ObjectReference `json:"relatedObjects"`
 }
 
+// RequiredRouteAnnotations specifies annotations that are required to be set on
+// newly created routes matching some criteria.
+type RequiredRouteAnnotations struct {
+	// domains is an optional list of domains for which these annotations
```
Comment (deads2k) [inline config/v1/types_ingress.go:186]: do we allow `*.department.foo.com`?

File: config/v1/types_ingress.go [hunk 3]
```diff
@@ -54,6 +54,14 @@ type IngressSpec struct {
 	// configurable routes.
 	// +optional
 	ComponentRoutes []ComponentRouteSpec `json:"componentRoutes,omitempty"`
+
+	// requiredRouteAnnotations is an optional list of default annotations
```
Comment (deads2k) [inline config/v1/types_ingress.go:58]: indicate how order is handled and how it can be used to exclude `openshift-*` namespaces.

File: config/v1/types_ingress.go (hunk 1 above)
Comment (deads2k) [inline config/v1/types_ingress.go:255]: if we narrow the scope to hsts, we can have options in here like 
1. max-age - required exact value, required no longer than, require no shorter than, no opinion
2. preload - required true, required false, no opinion
3. include-subdomains - required true, required false, no opinion

File: config/v1/types_ingress.go [hunk 4]
```diff
@@ -172,6 +180,98 @@ type ComponentRouteStatus struct {
...
+	// newly created routes.  The following values are allowed for this
+	// field:
+	//
+	// * "Allow" allows the user to provide arbitrary annotation values.
+	//   With this policy, if an annotation key in requiredAnnotations is
+	//   not specified on the route at all, the route is rejected.  If every
+	//   annotation key in requiredAnnotations is specified, then the route
+	//   is admitted, irrespective of the annotation values.
+	//
+	// * "AllowWithDefaulting" allows the user to provide arbitrary
```
Comment (deads2k) [inline config/v1/types_ingress.go:206]: Setting this as a default value on routes changes the meaning of a route depending on the cluster in which the route is created.  I think it's worth discussing with PM whether they really want to create non-replayable resources that vary based on configuration outside of the app owner's control and in some cases even outside the app owner's visibility.

I think that a cluster-admin goal of, "I want every route served for domain X to have HSTS" can be achieved without this mutation very quickly with a combination of a Deny and client-side `oc annotate route`

File: config/v1/types_ingress.go [hunk 5]
```diff
@@ -172,6 +180,98 @@ type ComponentRouteStatus struct {
...
+	//   requiredAnnotations.
+	//
+	// Note that the "AllowWithDefaulting" and "Override" options are
+	// dangerous to use.  With these options, a route that is created using
+	// a particular definition one cluster may behave differently from a
+	// route that is created using the same definition but on a cluster with
+	// different required route annotations configured.  Using "Allow" or
+	// "Deny" is safer because either option causes route creation to fail,
+	// loudly, if an annotation is not set as expected for the cluster.
+	Policy RequiredRouteAnnotationsPolicy `json:"policy"`
```
Comment (deads2k) [inline config/v1/types_ingress.go:232]: I think that if you narrow to just hsts as mentioned in comment https://github.com/openshift/api/pull/881/files#r605718112, you don't need to have a generic policy specified.  I think HSTS has a few useful dimensions to control with a few logical directions to have flexibility in

File: config/v1/types_ingress.go [hunk 6]
```diff
@@ -54,6 +54,14 @@ type IngressSpec struct {
 	// configurable routes.
 	// +optional
 	ComponentRoutes []ComponentRouteSpec `json:"componentRoutes,omitempty"`
+
+	// requiredRouteAnnotations is an optional list of default annotations
+	// for newly created routes. If requiredrouteAnnotations is nonempty,
+	// this value is used to generate default host values for Route.  Unlike
+	// domain, appsDomain may be modified after installation.  This assumes
+	// a new ingresscontroller has been setup with a wildcard certificate.
```
Comment (candita) [inline config/v1/types_ingress.go:62]: Are these two lines relevant to RequiredRouteAnnotations comments?

File: config/v1/types_ingress.go [hunk 7]
```diff
@@ -172,6 +180,98 @@ type ComponentRouteStatus struct {
...
+type RequiredRouteAnnotations struct {
+	// domains is an optional list of domains for which these annotations
+	// are required.  If domains is specified and a route is created with a
+	// spec.host matching one of the domains, the route must specify the
+	// annotations specified in requiredAnnotations.  If domains is empty,
+	// the specified annotations are required for all newly created routes.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Domains string `json:"domains,omitempty"`
```
Comment (candita) [inline config/v1/types_ingress.go:194]: I can foresee cases where we might not want to apply attributes to a list of `Domain`s.  For example, we might want to exclude namespacea from domaina, and namespaceb from domainb.  To express that we need to have something like a map keyed on `Domain`, with `Policy` and `NamespaceSelector` values for each domain rather than a list of domains with a single policy or namespace selector that has to be applied to all.

Comment (candita) [reply config/v1/types_ingress.go:194]: Or, the `IngressSpec.RequiredRouteAnnotations` expressed as a map with Domain as key?

**Comment by Miciah** [comment]
/close
Superseded by https://github.com/openshift/api/pull/958.
